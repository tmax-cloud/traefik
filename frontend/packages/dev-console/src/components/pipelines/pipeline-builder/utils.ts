import { history, resourcePathFromModel } from '@console/internal/components/utils';
import { apiVersionForModel } from '@console/internal/module/k8s';
import { ClusterTaskModel, PipelineModel } from '@console/internal/models';
import { Pipeline, PipelineResourceTask, PipelineResourceTaskParam, PipelineTask } from '../../../utils/pipeline-augment';
import { getTaskParameters } from '../resource-utils';
import { TASK_ERROR_STRINGS, TaskErrorType } from './const';
import { PipelineBuilderFormikValues, PipelineBuilderFormValues, TaskErrorMap } from './types';

export const getErrorMessage = (errorTypes: TaskErrorType[], errorMap: TaskErrorMap) => (taskName: string): string => {
  if (!taskName) {
    return TASK_ERROR_STRINGS[TaskErrorType.MISSING_NAME];
  }

  const errorList: TaskErrorType[] | undefined = errorMap?.[taskName];
  if (!errorList) return null;

  const hasRequestedError = errorList.filter(error => errorTypes.includes(error));
  return hasRequestedError.length > 0 ? TASK_ERROR_STRINGS[hasRequestedError[0]] : null;
};

export const taskParamIsRequired = (param: PipelineResourceTaskParam): boolean => {
  return false;
};

export const convertResourceToTask = (resource: PipelineResourceTask, runAfter?: string[]): PipelineTask => {
  return {
    name: resource.metadata.name,
    runAfter,
    taskRef: {
      kind: resource.kind === ClusterTaskModel.kind ? ClusterTaskModel.kind : undefined,
      name: resource.metadata.name,
    },
    params: getTaskParameters(resource).map(param => ({
      name: param.name,
      value: param.default,
      type: param.type
    })),
  };
};

export const getPipelineURL = (namespace: string) => {
  return `/k8s/ns/${namespace}/pipelines`;
};

const removeListRunAfters = (task: PipelineTask, listIds: string[]): PipelineTask => {
  if (task?.runAfter && listIds.length > 0) {
    // Trim out any runAfters pointing at list nodes
    const runAfter = (task.runAfter || []).filter(runAfterName => !listIds.includes(runAfterName));

    return {
      ...task,
      runAfter,
    };
  }

  return task;
};

const removeEmptyDefaultParams = (task: PipelineTask): PipelineTask => {
  if (task.params?.length > 0) {
    // Since we can submit, this param has a default; check for empty values and remove
    // 20210929 Policy changed
    task.params.forEach(param => { if (param.value === undefined) { param.value = (param.type && param.type === 'array') ?  [''] : ''; }; delete param.type; });
    return {
      ...task,
      //params: task.params.filter(param => !!param.value),
    };
  }

  return task;
};

export const convertBuilderFormToPipeline = (formValues: PipelineBuilderFormikValues, namespace: string, existingPipeline?: Pipeline): Pipeline => {
  const { name, resources, params, workspaces, tasks, listTasks, metadata } = formValues;
  const listIds = listTasks.map(listTask => listTask.name);
  const labels = metadata?.labels;

  return {
    ...existingPipeline,
    apiVersion: apiVersionForModel(PipelineModel),
    kind: PipelineModel.kind,
    metadata: {
      ...existingPipeline?.metadata,
      name,
      namespace,
      labels,
    },
    spec: {
      ...existingPipeline?.spec,
      params,
      resources,
      workspaces,
      tasks: tasks.map(task => removeEmptyDefaultParams(removeListRunAfters(task, listIds))),
    },
  };
};

export const convertPipelineToBuilderForm = (pipeline: Pipeline): PipelineBuilderFormValues => {
  if (!pipeline) return null;

  const {
    metadata: { name },
    spec: { params = [], resources = [], workspaces = [], tasks = [] },
  } = pipeline;

  return {
    name,
    params,
    resources,
    workspaces,
    tasks,
    listTasks: [],
  };
};

export const goToYAML = (existingPipeline?: Pipeline, namespace?: string) => {
  history.push(existingPipeline ? `${resourcePathFromModel(PipelineModel, existingPipeline?.metadata?.name, existingPipeline?.metadata?.namespace)}/yaml` : `${getPipelineURL(namespace)}/~new`);
};
