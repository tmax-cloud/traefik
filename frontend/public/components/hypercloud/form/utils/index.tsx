export * from './modal-launcher';
export * from './modal-list';
