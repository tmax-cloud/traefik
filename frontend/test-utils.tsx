import * as React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from './jest-setup';

// MEMO : redux Provider로 감싼 구조로 @testing-library/react의 render함수를 래핑함.
const render = (ui: React.ReactElement, options?) => {
  const Wrapper = ({ children }: { children?: React.ReactNode }) => {
    return <Provider store={store}>{!!options?.wrapper ? options.wrapper({ children: children }) : children}</Provider>;
  };

  // MEMO : options로 들어온 wrapper는 Wrapper컴포넌트에서 사용하고, 여기서 wrapper속성을 Wrapper로 덮어씌워서 설정해 줌.
  return rtlRender(ui, { ...options, wrapper: Wrapper });
};

export * from '@testing-library/react';
export { render };
