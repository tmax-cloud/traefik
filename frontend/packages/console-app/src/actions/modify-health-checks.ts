import * as _ from 'lodash';
import { K8sKind, K8sResourceKind, referenceFor } from '@console/internal/module/k8s';
import { KebabOption } from '@console/internal/components/utils';

const healthChecksAdded = (resource: K8sResourceKind): boolean => {
  const containers = resource?.spec?.template?.spec?.containers;
  return _.every(containers, container => container.readinessProbe || container.livenessProbe || container.startupProbe);
};

const healthChecksUrl = (model: K8sKind, obj: K8sResourceKind): string => {
  const {
    metadata: { name, namespace },
  } = obj;
  const resourcePlural = model.crd ? referenceFor(obj) : model.plural;
  const containers = obj?.spec?.template?.spec?.containers;
  const containerName = containers?.[0]?.name;
  return `/k8s/ns/${namespace}/${resourcePlural}/${name}/containers/${containerName}/health-checks`;
};

export const AddHealthChecks = (model: K8sKind, obj: K8sResourceKind): KebabOption => {
  return {
    label: 'COMMON:MSG_MAIN_ACTIONBUTTON_14',
    hidden: healthChecksAdded(obj),
    href: healthChecksUrl(model, obj),
    accessReview: {
      group: model.apiGroup,
      resource: model.plural,
      name: obj.metadata.name,
      namespace: obj.metadata.namespace,
      verb: 'update',
    },
  };
};

export const EditHealthChecks = (model: K8sKind, obj: K8sResourceKind): KebabOption => {
  return {
    label: 'COMMON:MSG_MAIN_ACTIONBUTTON_9',
    hidden: !healthChecksAdded(obj),
    href: healthChecksUrl(model, obj),
    accessReview: {
      group: model.apiGroup,
      resource: model.plural,
      name: obj.metadata.name,
      namespace: obj.metadata.namespace,
      verb: 'update',
    },
  };
};
