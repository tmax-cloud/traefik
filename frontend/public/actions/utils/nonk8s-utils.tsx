import * as _ from 'lodash-es';
import { CustomMenusMap } from '@console/internal/hypercloud/menu/menu-types';
import { getIngressUrl } from '@console/internal/components/hypercloud/utils/ingress-utils';
import { coFetchJSON } from '../../co-fetch';
import { history } from '@console/internal/components/utils/router';

export const getHelmHost = async () => {
  const mapUrl = (CustomMenusMap as any).Helm.url;
  return mapUrl !== '' ? mapUrl : await getIngressUrl('helm-apiserver');
};

export const helmAPI = '/api/kubernetes/apis/helmapi.tmax.io/v1';

export const getKind = (id: string) => {
  return id.split('~~')[0];
};
const getHelmRepo = (id: string) => {
  return id.split('~~')[1];
};

//get object api url 반환
export const nonK8sObjectUrl = async (id: string, namespace: string, name: string) => {
  const kind = getKind(id);
  let host = '';

  switch (kind) {
    case 'HelmRepository':
      return `${helmAPI}/repos/${name}`;
    case 'HelmRelease':
      host = await getHelmHost();
      return `${host}/helm/v1/namespaces/${namespace}/releases/${name}`;
    case 'HelmChart':
      host = await getHelmHost();
      const helmRepo = getHelmRepo(id);
      return `${host}/helm/v1/charts/${helmRepo}_${name}`;
    default:
      return '';
  }
};

//get object 결과 정리
export const nonK8sObjectResult = (kind: string, response: any) => {
  switch (kind) {
    case 'HelmRepository':
      return response.repoInfo[0];
    case 'HelmRelease':
      return response.release[0];
    case 'HelmChart':
      let entriesvalues = Object.values(_.get(response, 'indexfile.entries'));
      return entriesvalues[0][0];
    default:
      return {};
  }
};

//get list api url 반환
export const nonK8sListUrl = async (id: string, query: any) => {
  const kind = getKind(id);
  let host = '';

  switch (kind) {
    case 'HelmRepository':
      return `${helmAPI}/repos`;
    case 'HelmRelease':
      host = await getHelmHost();
      return query?.ns ? `${host}/helm/v1/namespaces/${query.ns}/releases` : `${host}/helm/v1/releases`;
    case 'HelmChart':
      host = await getHelmHost();
      const helmRepo = getHelmRepo(id);
      return helmRepo ? `${host}/helm/v1/charts?repository=${helmRepo}` : `${host}/helm/v1/charts`;
    default:
      return '';
  }
};
//get list 결과 정리
export const nonK8sListResult = async (id: string, response: any) => {
  const kind = getKind(id);
  switch (kind) {
    case 'HelmRepository':
      if (response.error) {
        return [];
      }
      await (async () => {
        await Promise.all(
          response.repoInfo.map(async repoinfo => {
            const helmHost = await getHelmHost();
            if (helmHost === null) {
              history.push('/ingress-check?ingresslabelvalue=helm-apiserver');
            }
            const response = await coFetchJSON(`${helmHost}/helm/v1/charts?repository=${repoinfo.name}`);
            let tempList = [];
            let entriesvalues = Object.values(_.get(response, 'indexfile.entries'));
            entriesvalues.map(value => {
              tempList.push(value[0]);
            });
            repoinfo.charts = tempList;
          }),
        );
      })();
      return response.repoInfo;
    case 'HelmRelease':
      if (response.error) {
        return [];
      }
      return response.release;
    case 'HelmChart':
      if (response.error) {
        return [];
      }
      let tempList = [];
      let entriesvalues = Object.values(_.get(response, 'indexfile.entries'));
      entriesvalues.map(value => {
        tempList.push(value[0]);
      });
      return tempList;
    default:
      return [];
  }
};
