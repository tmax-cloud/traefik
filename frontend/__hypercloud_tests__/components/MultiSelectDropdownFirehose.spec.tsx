import * as React from 'react';
import { MultiSelectDropdownFirehose, MultiSelectDropdownFirehoseProps } from '@console/internal/components/hypercloud/utils/multi-dropdown-new';
import { render } from '../../test-utils';
import { DeploymentModel } from '@console/internal/models';
import { DeploymentsTestData } from '../../__hypercloud_mocks__/firehoseMocks';
import * as _ from 'lodash-es';
import store from '@console/internal/redux';
import userEvent from '@testing-library/user-event';
import { useForm, FormProvider, Controller } from 'react-hook-form';
import * as k8sActions from '@console/internal/actions/k8s';
import { inject } from '@console/internal/components/utils/inject';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: str => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
  withTranslation: () => Component => {
    Component.defaultProps = { ...Component.defaultProps, t: () => '' };
    return Component;
  },
}));

jest.mock('@console/internal/components/utils/firehose', () => {
  return {
    Firehose: props => {
      return inject(props.children, {
        filters: null,
        loaded: true,
        loadError: null,
        resources: {
          deployments: DeploymentsTestData,
        },
      });
    },
  };
});

let getValues;

const renderMultiSelectDropdownFirehoseForm = (props: MultiSelectDropdownFirehoseProps) => {
  return render(<MultiSelectDropdownFirehose {...props} />, {
    wrapper: ({ children }) => {
      const methods = useForm();
      getValues = methods.getValues;
      return (
        <FormProvider {...methods}>
          <form>
            <Controller
              as={children}
              control={methods.control}
              name={props.name}
              onChange={([selected]) => {
                return { value: selected };
              }}
            />
            <button type="submit">Submit</button>
          </form>
        </FormProvider>
      );
    },
  });
};

describe('ResourceDropdown Test', () => {
  beforeEach(() => {
    const resources = {
      models: [DeploymentModel],
      adminResources: [],
      allResources: [],
      configResources: [],
      namespacedSet: null,
      safeResources: [],
      preferredVersions: [],
    };
    const action = k8sActions.receivedResources(resources);
    store.dispatch(action);
  });

  it('드롭다운 하위 모든 아이템을 선택하면 All chip이 추가되고 value도 세팅돼야 합니다.', async () => {
    const { getByText, getAllByText } = await renderMultiSelectDropdownFirehoseForm({
      name: 'MultiDropdownFirehose',
      resourcesConfig: [{ kind: 'Deployment', prop: 'deployment', isList: true }],
      kind: 'Deployment',
      shrinkOnSelectAll: true,
      selectAllChipObj: { label: 'All Select', value: 'all' },
    });
    userEvent.click(getByText('Select Resources'));
    userEvent.click(getAllByText('Deployment')[0]);
    userEvent.click(getAllByText('Deployment')[1]);
    userEvent.click(getByText('Select Resources'));
    getByText('2');
    getByText('All Select');
    expect(getValues?.()).toEqual({ MultiDropdownFirehose: { label: 'All Select', value: 'all' } });
  });

  it('모든 아이템을 체크한 상태에서 All의 체크를 해제하면 value가 비어야 합니다.', async () => {
    const { container, getByText, getAllByText, getAllByRole } = await renderMultiSelectDropdownFirehoseForm({
      name: 'MultiDropdownFirehose2',
      resourcesConfig: [{ kind: 'Deployment', prop: 'deployment', isList: true }],
      kind: 'Deployment',
      shrinkOnSelectAll: true,
      selectAllChipObj: { label: 'All Select', value: 'all' },
    });
    userEvent.click(getByText('Select Resources'));
    userEvent.click(getAllByText('Deployment')[0]);
    userEvent.click(getAllByText('Deployment')[1]);
    expect(getAllByRole('checkbox')[0]).toBeChecked();
    userEvent.click(getAllByText('All')[1]);
    expect(getValues?.()).toEqual({ MultiDropdownFirehose2: [] });
    expect(container).toMatchSnapshot();
  });

  it('검색 필터링 툴바 테스트입니다.', async () => {
    const { container, getByText, getByRole } = await renderMultiSelectDropdownFirehoseForm({
      name: 'MultiDropdownFirehose3',
      resourcesConfig: [{ kind: 'Deployment', prop: 'deployment', isList: true }],
      kind: 'Deployment',
      selectAllChipObj: { label: 'All Select', value: 'all' },
    });
    userEvent.click(getByText('Select Resources'));
    userEvent.type(getByRole('textbox'), 'catalog-catalog-w');
    getByText('catalog-catalog-webhook');
    expect(container).toMatchSnapshot();
  });

  it('shrinkOnSelectAll가 false인 드롭다운의 모든 아이템 선택 후 clear all 클릭 동작 테스트입니다.', async () => {
    const { container, getByText, getAllByText } = await renderMultiSelectDropdownFirehoseForm({
      name: 'MultiDropdownFirehose4',
      resourcesConfig: [{ kind: 'Deployment', prop: 'deployment', isList: true }],
      kind: 'Deployment',
      shrinkOnSelectAll: false,
    });
    userEvent.click(getByText('Select Resources'));
    userEvent.click(getAllByText('Deployment')[0]);
    userEvent.click(getAllByText('Deployment')[1]);
    expect(getValues?.()).toEqual({
      MultiDropdownFirehose4: [
        {
          key: 'Deployment_catalog-catalog-controller-manager',
          apiVersion: '',
          kind: 'Deployment',
          label: 'catalog-catalog-controller-manager',
          value: 'catalog-catalog-controller-manager',
        },
        {
          key: 'Deployment_catalog-catalog-webhook',
          apiVersion: '',
          kind: 'Deployment',
          label: 'catalog-catalog-webhook',
          value: 'catalog-catalog-webhook',
        },
      ],
    });
    userEvent.click(getAllByText('Clear all')[1]);
    expect(getValues?.()).toEqual({ MultiDropdownFirehose4: [] });
    expect(container).toMatchSnapshot();
  });
});
