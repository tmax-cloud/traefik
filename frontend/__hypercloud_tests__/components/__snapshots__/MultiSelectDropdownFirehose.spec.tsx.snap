// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceDropdown Test shrinkOnSelectAll가 false인 드롭다운의 모든 아이템 선택 후 clear all 클릭 동작 테스트입니다. 1`] = `
<div>
  <form>
    <div
      class="pf-c-data-toolbar"
      id="multidropdown-toolbar"
    >
      <div
        class="pf-c-data-toolbar__content"
      >
        <div
          class="pf-c-data-toolbar__content-section"
        >
          <div
            class="pf-c-data-toolbar__item"
          >
            <div
              class="pf-c-data-toolbar__item"
            >
              <div
                style="position: relative;"
              >
                <div
                  class="hc-dropdown-main-button"
                  style="width: 200px;"
                >
                  <span
                    style="margin: 0px 3px; text-overflow: ellipsis; display: block; white-space: nowrap; overflow-x: hidden; width: calc(100% - 60px);"
                  >
                    Select Resources
                  </span>
                  <span
                    class="pf-c-badge pf-m-read"
                  >
                    0
                  </span>
                  <svg
                    aria-hidden="true"
                    class="hc-dropdown-select_toggle-icon"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="margin-left: 10px; position: absolute; right: 0px;"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                      transform=""
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="pf-c-data-toolbar__expandable-content"
          id="multidropdown-toolbar-expandable-content-26"
        >
          <div
            class="pf-c-data-toolbar__group"
          />
        </div>
      </div>
      <div
        class="pf-c-data-toolbar__content pf-m-hidden"
        hidden=""
      >
        <div
          class="pf-c-data-toolbar__group"
        />
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`ResourceDropdown Test 검색 필터링 툴바 테스트입니다. 1`] = `
<div>
  <form>
    <div
      class="pf-c-data-toolbar"
      id="multidropdown-toolbar"
    >
      <div
        class="pf-c-data-toolbar__content"
      >
        <div
          class="pf-c-data-toolbar__content-section"
        >
          <div
            class="pf-c-data-toolbar__item"
          >
            <div
              class="pf-c-data-toolbar__item"
            >
              <div
                style="position: relative;"
              >
                <div
                  class="hc-dropdown-main-button"
                  style="width: 200px;"
                >
                  <span
                    style="margin: 0px 3px; text-overflow: ellipsis; display: block; white-space: nowrap; overflow-x: hidden; width: calc(100% - 60px);"
                  >
                    Select Resources
                  </span>
                  <span
                    class="pf-c-badge pf-m-read"
                  >
                    0
                  </span>
                  <svg
                    aria-hidden="true"
                    class="hc-dropdown-select_toggle-icon"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="margin-left: 10px; position: absolute; right: 0px;"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                      transform=""
                    />
                  </svg>
                </div>
                <div
                  style="background-color: white; border-radius: 4px; box-shadow: 0 0 0 1px hsla(218, 50%, 10%, 0.1), 0 4px 11px hsla(218, 50%, 10%, 0.1); margin-top: 2px; position: absolute; z-index: 2; overflow-y: auto;"
                >
                  <div
                    class=" css-1xz6ynz-container"
                  >
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    >
                      <span
                        id="aria-selection"
                      />
                      <span
                        id="aria-context"
                      >
                        option catalog-catalog-webhook focused, 3 of 3. 1 result available for search term catalog-catalog-w. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu.
                      </span>
                    </span>
                    <div
                      class="hc-select__control hc-select__control--is-focused hc-select__control--menu-is-open css-11hhl9n-control"
                    >
                      <div
                        class="hc-select__value-container hc-select__value-container--is-multi css-g1d714-ValueContainer"
                      >
                        <div
                          class="css-b8ldur-Input"
                        >
                          <div
                            class="hc-select__input"
                            style="display: inline-block;"
                          >
                            <input
                              aria-autocomplete="list"
                              autocapitalize="none"
                              autocomplete="off"
                              autocorrect="off"
                              id="react-select-4-input"
                              spellcheck="false"
                              style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                              tabindex="0"
                              type="text"
                              value="catalog-catalog-w"
                            />
                            <div
                              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                            >
                              catalog-catalog-w
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="hc-select__indicators css-1hb7zxy-IndicatorsContainer"
                      />
                    </div>
                    <div
                      class="hc-select__menu css-aey3sy-Menu"
                    >
                      <div
                        class="hc-select__menu-list hc-select__menu-list--is-multi css-4ljt47-MenuList"
                      >
                        <div
                          class="hc-select__option hc-select__option--is-focused css-1nuv1fz-option"
                          id="react-select-4-option-2"
                          tabindex="-1"
                        >
                          <span
                            class="co-resource-item"
                            id="hc-multidropdown-item"
                          >
                            <input
                              id="hc-multidropdown-item"
                              style="margin-right: 3px;"
                              type="checkbox"
                            />
                            <span
                              class="co-resource-icon--fixed-width"
                              id="hc-multidropdown-item"
                            >
                              <span
                                class="sr-only"
                              >
                                Deployment
                              </span>
                              <span
                                class="co-m-resource-icon co-m-resource-deployment"
                                title="Deployment"
                              >
                                D
                              </span>
                            </span>
                            <span
                              class="co-resource-item__resource-name"
                              id="hc-multidropdown-item"
                              style="margin: 0px 3px; text-overflow: ellipsis; display: block; white-space: nowrap; overflow-x: hidden;"
                            >
                              <span
                                id="hc-multidropdown-item"
                              >
                                catalog-catalog-webhook
                              </span>
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                    <div>
                      <input
                        name="MultiDropdownFirehose3"
                        type="hidden"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="pf-c-data-toolbar__expandable-content"
          id="multidropdown-toolbar-expandable-content-17"
        >
          <div
            class="pf-c-data-toolbar__group"
          />
        </div>
      </div>
      <div
        class="pf-c-data-toolbar__content pf-m-hidden"
        hidden=""
      >
        <div
          class="pf-c-data-toolbar__group"
        />
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`ResourceDropdown Test 모든 아이템을 체크한 상태에서 All의 체크를 해제하면 value가 비어야 합니다. 1`] = `
<div>
  <form>
    <div
      class="pf-c-data-toolbar"
      id="multidropdown-toolbar"
    >
      <div
        class="pf-c-data-toolbar__content"
      >
        <div
          class="pf-c-data-toolbar__content-section"
        >
          <div
            class="pf-c-data-toolbar__item"
          >
            <div
              class="pf-c-data-toolbar__item"
            >
              <div
                style="position: relative;"
              >
                <div
                  class="hc-dropdown-main-button"
                  style="width: 200px;"
                >
                  <span
                    style="margin: 0px 3px; text-overflow: ellipsis; display: block; white-space: nowrap; overflow-x: hidden; width: calc(100% - 60px);"
                  >
                    Select Resources
                  </span>
                  <span
                    class="pf-c-badge pf-m-read"
                  >
                    0
                  </span>
                  <svg
                    aria-hidden="true"
                    class="hc-dropdown-select_toggle-icon"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="margin-left: 10px; position: absolute; right: 0px;"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                      transform=""
                    />
                  </svg>
                </div>
                <div
                  style="background-color: white; border-radius: 4px; box-shadow: 0 0 0 1px hsla(218, 50%, 10%, 0.1), 0 4px 11px hsla(218, 50%, 10%, 0.1); margin-top: 2px; position: absolute; z-index: 2; overflow-y: auto;"
                >
                  <div
                    class=" css-1xz6ynz-container"
                  >
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="hc-select__control hc-select__control--menu-is-open css-odw0t-control"
                    >
                      <div
                        class="hc-select__value-container hc-select__value-container--is-multi css-g1d714-ValueContainer"
                      >
                        <div
                          class="hc-select__placeholder css-1wa3eu0-placeholder"
                        >
                          Select...
                        </div>
                        <div
                          class="css-b8ldur-Input"
                        >
                          <div
                            class="hc-select__input"
                            style="display: inline-block;"
                          >
                            <input
                              aria-autocomplete="list"
                              autocapitalize="none"
                              autocomplete="off"
                              autocorrect="off"
                              id="react-select-3-input"
                              spellcheck="false"
                              style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                              tabindex="0"
                              type="text"
                              value=""
                            />
                            <div
                              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="hc-select__indicators css-1hb7zxy-IndicatorsContainer"
                      />
                    </div>
                    <div
                      class="hc-select__menu css-aey3sy-Menu"
                    >
                      <div
                        class="hc-select__menu-list hc-select__menu-list--is-multi css-4ljt47-MenuList"
                      >
                        <div
                          class="hc-select__option hc-select__option--is-focused css-1nuv1fz-option"
                          id="react-select-3-option-0"
                          tabindex="-1"
                        >
                          <span
                            class="co-resource-item"
                            id="hc-multidropdown-item"
                          >
                            <input
                              id="hc-multidropdown-item"
                              style="margin-right: 3px;"
                              type="checkbox"
                            />
                            <span
                              class="co-resource-icon--fixed-width"
                              id="hc-multidropdown-item"
                            >
                              <span
                                class="sr-only"
                              >
                                All
                              </span>
                              <span
                                class="co-m-resource-icon co-m-resource-all"
                                title="All"
                              >
                                A
                              </span>
                            </span>
                            <span
                              class="co-resource-item__resource-name"
                              id="hc-multidropdown-item"
                              style="margin: 0px 3px; text-overflow: ellipsis; display: block; white-space: nowrap; overflow-x: hidden;"
                            >
                              <span
                                id="hc-multidropdown-item"
                              >
                                All
                              </span>
                            </span>
                          </span>
                        </div>
                        <div
                          class="hc-select__option css-1nuv1fz-option"
                          id="react-select-3-option-1"
                          tabindex="-1"
                        >
                          <span
                            class="co-resource-item"
                            id="hc-multidropdown-item"
                          >
                            <input
                              id="hc-multidropdown-item"
                              style="margin-right: 3px;"
                              type="checkbox"
                            />
                            <span
                              class="co-resource-icon--fixed-width"
                              id="hc-multidropdown-item"
                            >
                              <span
                                class="sr-only"
                              >
                                Deployment
                              </span>
                              <span
                                class="co-m-resource-icon co-m-resource-deployment"
                                title="Deployment"
                              >
                                D
                              </span>
                            </span>
                            <span
                              class="co-resource-item__resource-name"
                              id="hc-multidropdown-item"
                              style="margin: 0px 3px; text-overflow: ellipsis; display: block; white-space: nowrap; overflow-x: hidden;"
                            >
                              <span
                                id="hc-multidropdown-item"
                              >
                                catalog-catalog-controller-manager
                              </span>
                            </span>
                          </span>
                        </div>
                        <div
                          class="hc-select__option css-1nuv1fz-option"
                          id="react-select-3-option-2"
                          tabindex="-1"
                        >
                          <span
                            class="co-resource-item"
                            id="hc-multidropdown-item"
                          >
                            <input
                              id="hc-multidropdown-item"
                              style="margin-right: 3px;"
                              type="checkbox"
                            />
                            <span
                              class="co-resource-icon--fixed-width"
                              id="hc-multidropdown-item"
                            >
                              <span
                                class="sr-only"
                              >
                                Deployment
                              </span>
                              <span
                                class="co-m-resource-icon co-m-resource-deployment"
                                title="Deployment"
                              >
                                D
                              </span>
                            </span>
                            <span
                              class="co-resource-item__resource-name"
                              id="hc-multidropdown-item"
                              style="margin: 0px 3px; text-overflow: ellipsis; display: block; white-space: nowrap; overflow-x: hidden;"
                            >
                              <span
                                id="hc-multidropdown-item"
                              >
                                catalog-catalog-webhook
                              </span>
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                    <div>
                      <input
                        name="MultiDropdownFirehose2"
                        type="hidden"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="pf-c-data-toolbar__expandable-content"
          id="multidropdown-toolbar-expandable-content-14"
        >
          <div
            class="pf-c-data-toolbar__group"
          />
        </div>
      </div>
      <div
        class="pf-c-data-toolbar__content pf-m-hidden"
        hidden=""
      >
        <div
          class="pf-c-data-toolbar__group"
        />
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
